openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go

```diff
@@ -142,3 +144,94 @@ func validateTolerations(versionedTolerations []corev1.Toleration, fldPath *fiel
 	allErrors = append(allErrors, apivalidation.ValidateTolerations(unversionedTolerations, fldPath)...)
 	return allErrors
 }
+
+// validateUpstreamResolversCreate validates configuration of the Upstream objects when TLS is configured.
+func validateUpstreamResolversCreate(upstreamResolvers operatorv1.UpstreamResolvers, fieldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	errs = append(errs, validateDNSTransportConfig(upstreamResolvers.TransportConfig, fieldPath)...)
+
+	switch upstreamResolvers.TransportConfig.Transport {
+	case "", operatorv1.CleartextTransport:
+		// Transport is empty or Cleartext so mixed Upstream types are allowed.
+		return errs
+	case operatorv1.TLSTransport:
+		// Transport is TLS so we must check if there are mixed Upstream types. SystemResolveConf is not allowed with TLS.
+		for i, upstream := range upstreamResolvers.Upstreams {
+			errPath := fieldPath.Child("upstreams")
+			if upstream.Type == operatorv1.SystemResolveConfType {
+				errMessage := fmt.Sprintf("upstream has type %s but %s has been configured", upstream.Type, operatorv1.TLSTransport)
+				errs = append(errs, field.Invalid(errPath.Index(i).Child("type"), upstream.Type, errMessage))
+			}
+		}
+	default:
+		// An unrecognized value has been configured for Transport so we'll need to append an error.
+		errMessage := fmt.Sprintf("upstreamResolvers has an unknown transport type %s", upstreamResolvers.TransportConfig.Transport)
+		errs = append(errs, field.Invalid(fieldPath.Child("transportConfig").Child("transport"), upstreamResolvers.TransportConfig.Transport, errMessage))
+	}
+
+	return errs
+}
+
+// validateUpstreamResolversUpdate validates configuration of the Upstream objects when TLS is configured.
+func validateUpstreamResolversUpdate(newUpstreamResolvers operatorv1.UpstreamResolvers, oldUpstreamResolvers operatorv1.UpstreamResolvers, fieldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	// On update, return early if the same Transport is set.
```

Comment:
why do we return early from all validation if other stanzas are changed?  Split validation can make this reasonable, but please structure it so that added stanzas clearly need to be checked.

```go
if stanzaOne is not the same{
  errs = append(check the errors)
}
```

this way copy/paste will work in your favor as you add stanzas.

Reply:
That's excellent feedback. I've updated both `validateUpstreamResolversUpdate` and `validateServersUpdate` using this suggestion in https://github.com/openshift/kubernetes/pull/1247/commits/56e56a13b3410d412a24b54dc987070202451149